testcases/crypto/aes_func.c

```diff
@@ -1929,6 +2006,308 @@ CK_RV do_SignVerifyMAC(struct published_mac_test_suite_info *tsuite)
+        goto testcase_cleanup;
+    }
+
+    /* Encrypt some data with this key object */
+    rc = funcs->C_EncryptInit(session, &encr_mech, h_key);
+    if (rc != CKR_OK) {
+        testcase_fail("C_EncryptInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    memset(output, 0, sizeof(output));
+    rc = funcs->C_Encrypt(session, input, sizeof(input), output, &outlen);
+    if (rc != CKR_OK) {
+        testcase_fail("C_Encrypt rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    testcase_pass("Generate token key object and encrypt");
+    testcase_begin("Log out, close session, and login in again in new session");
+    testcase_new_assertion();
+
+    /* Log out and close session */
+    testcase_user_logout();
+    rc = funcs->C_CloseAllSessions(slot_id);
+    if (rc != CKR_OK) {
+        testcase_fail("C_CloseAllSessions rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    /* Log in again read/only, ok for just decrypt with an existing pkey */
+    testcase_rw_session();
+    testcase_user_login();
+
+    testcase_pass("Log out, close session, and login in again in new session");
+    testcase_begin("Retrieve the key from repository");
+    testcase_new_assertion();
+
+    /* Retrieve the key from repository */
+    rc = funcs->C_FindObjectsInit(session, find_tmpl, find_tmpl_len);
+    if (rc != CKR_OK) {
+        testcase_fail("C_FindObjectsInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    rc = funcs->C_FindObjects(session, &h_key2, 2, &count);
+    if (rc != CKR_OK) {
+        testcase_fail("C_FindObjects rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+    if (count == 0) {
+        testcase_fail("Didn't find key with label '%s'in repository", label);
+        goto testcase_cleanup;
+    } else if (count > 1) {
+        testcase_skip("Found %ld objs with label '%s', only expected one. "
+                      "Skipping this test.\n", count, label);
+        rc = funcs->C_FindObjectsFinal(session);
+        goto testcase_cleanup;
+    }
+
+    /* h_key2 points to the same object as h_key in our first session, so
+     * invalidate h_key to avoid a double destroy obj later. */
+    h_key = CK_INVALID_HANDLE;
+
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        testcase_fail("C_FindObjectsFinal rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    testcase_pass("Retrieve the key from repository");
+    testcase_begin("Decrypt encrypted data with retrieved token object");
+    testcase_new_assertion();
+
+    /* Decrypt encrypted data with the retrieved token object */
+    rc = funcs->C_DecryptInit(session, &encr_mech, h_key2);
+    if (rc != CKR_OK) {
+        testcase_fail("C_DecryptInit rc=%s", p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+
+    memset(decrypted, 0, sizeof(decrypted));
+    rc = funcs->C_Decrypt(session, output, sizeof(input), decrypted, &decrlen);
```

Comment:
decrlen must be initialized (to sizeof(decrypted)).

Reply:
Done.